### PR TITLE
Fix field evaluation in scaffolding

### PR DESCRIPTION
**Pull request by bgrauer-atacom**
- #1131
- #1134 (also removing invalid empty space prefix in CHAR evaluation)

### CODE REVIEW

**Comment by bgrauer-atacom** [comment]
Well, it was not my intention to put two issues into the same pull request.
Let me know if you prefer two seperate pull requests.

**Comment by cincuranet** [comment]
Single is fine.

**Comment by bgrauer-atacom** [comment]
Fixed invalid test table length.

File: src/FirebirdSql.EntityFrameworkCore.Firebird.Tests/Scaffolding/ScaffoldingTests.cs [hunk 1]
```diff
@@ -15,21 +15,117 @@
...
 using FirebirdSql.EntityFrameworkCore.Firebird.Scaffolding.Internal;
+using Microsoft.EntityFrameworkCore;
 using Microsoft.EntityFrameworkCore.Scaffolding;
 using NUnit.Framework;
 
 namespace FirebirdSql.EntityFrameworkCore.Firebird.Tests.Scaffolding;
 #pragma warning disable EF1001
 public class ScaffoldingTests : EntityFrameworkCoreTestsBase
 {
+	class ScaffoldingContext : FbTestDbContext
```
Comment (cincuranet) [inline src/FirebirdSql.EntityFrameworkCore.Firebird.Tests/Scaffolding/ScaffoldingTests.cs:29]: Where is this class used?

File: src/FirebirdSql.EntityFrameworkCore.Firebird.Tests/Scaffolding/ScaffoldingTests.cs [hunk 2]
```diff
@@ -15,21 +15,117 @@
...
+		// Arrange
+		var tableName = "TEST_READS_IS_NULL_FROM_DOMAIN";
+		var columnNameNoDomainNull = "NO_DOMAIN_NULL";
+		var columnNameNoDomainNotNull = "NO_DOMAIN_NOT_NULL";
+		var columnNameDomainNull = "DOMAIN_NULL";
+		var columnNameDomainNotNull = "DOMAIN_NOT_NUL";
+
+		using var commandDomainNull = Connection.CreateCommand();
+		commandDomainNull.CommandText = "create domain DOMAIN_NULL as INTEGER";
+		await commandDomainNull.ExecuteNonQueryAsync();
```
Comment (cincuranet) [inline src/FirebirdSql.EntityFrameworkCore.Firebird.Tests/Scaffolding/ScaffoldingTests.cs:60]: All database objects created should be removed in `TearDown`.

File: src/FirebirdSql.EntityFrameworkCore.Firebird.Tests/Scaffolding/ScaffoldingTests.cs [hunk 3]
```diff
@@ -15,21 +15,117 @@
...
+
+		using var commandDomainNotNull = Connection.CreateCommand();
+		commandDomainNotNull.CommandText = "create domain DOMAIN_NOT_NULL as INTEGER NOT NULL";
+		await commandDomainNotNull.ExecuteNonQueryAsync();
```
Comment (cincuranet) [inline src/FirebirdSql.EntityFrameworkCore.Firebird.Tests/Scaffolding/ScaffoldingTests.cs:64]: All database objects created should be removed in `TearDown`.

File: src/FirebirdSql.EntityFrameworkCore.Firebird.Tests/Scaffolding/ScaffoldingTests.cs [hunk 4]
```diff
@@ -15,21 +15,117 @@
...
+
+		using var commandTable = Connection.CreateCommand();
+		commandTable.CommandText = $"create table {tableName} ({columnNameNoDomainNull} INTEGER, {columnNameNoDomainNotNull} INTEGER NOT NULL, {columnNameDomainNull} DOMAIN_NULL, {columnNameDomainNotNull} DOMAIN_NOT_NULL)";
+		await commandTable.ExecuteNonQueryAsync();
```
Comment (cincuranet) [inline src/FirebirdSql.EntityFrameworkCore.Firebird.Tests/Scaffolding/ScaffoldingTests.cs:68]: All database objects created should be removed in `TearDown`.

File: src/FirebirdSql.EntityFrameworkCore.Firebird.Tests/Scaffolding/ScaffoldingTests.cs [hunk 5]
```diff
@@ -15,21 +15,117 @@
...
+	[TestCase("NUMERIC(18,1)", "NUMERIC_18_1")]
+	public async Task ReadsCorrectFieldType(string dataType, string tableSuffix)
+	{
+		// Arrange
+		var tableName = $"TEST_READS_CORRECT_{tableSuffix}";
+		var columnName = "FIELD";
+
+		using var commandTable = Connection.CreateCommand();
+		commandTable.CommandText = $"create table {tableName} ({columnName} {dataType})";
+		await commandTable.ExecuteNonQueryAsync();
```
Comment (cincuranet) [inline src/FirebirdSql.EntityFrameworkCore.Firebird.Tests/Scaffolding/ScaffoldingTests.cs:117]: All database objects created should be removed in `TearDown`.

**Review by cincuranet — CHANGES_REQUESTED** [review]

File: src/FirebirdSql.EntityFrameworkCore.Firebird.Tests/Scaffolding/ScaffoldingTests.cs (hunk 4 above)
Comment (bgrauer-atacom) [reply src/FirebirdSql.EntityFrameworkCore.Firebird.Tests/Scaffolding/ScaffoldingTests.cs:68]: Going thru other tests:
- [QueryTests](https://github.com/FirebirdSQL/NETProvider/blob/f169563ec05067c55974d2dfe09ff4f64cfd05d9/src/EntityFramework.Firebird.Tests/QueryTests.cs#L50) / [InsertTests](https://github.com/FirebirdSQL/NETProvider/blob/f169563ec05067c55974d2dfe09ff4f64cfd05d9/src/FirebirdSql.EntityFrameworkCore.Firebird.Tests/EndToEnd/InsertTests.cs#L53) and [UpdateTests](https://github.com/FirebirdSQL/NETProvider/blob/f169563ec05067c55974d2dfe09ff4f64cfd05d9/src/FirebirdSql.EntityFrameworkCore.Firebird.Tests/EndToEnd/UpdateTests.cs#L56) do not remove its tables (or I missed out the important point)
- [FbBatchCommandTests](https://github.com/FirebirdSQL/NETProvider/blob/f169563ec05067c55974d2dfe09ff4f64cfd05d9/src/FirebirdSql.Data.FirebirdClient.Tests/FbBatchCommandTests.cs#L50) creates tables in SetUp and removes them in TearDown
- [FbBatchCommandTests](https://github.com/FirebirdSQL/NETProvider/blob/f169563ec05067c55974d2dfe09ff4f64cfd05d9/src/FirebirdSql.Data.FirebirdClient.Tests/FbCommandTests.cs#L260) uses in some cases try / finally 

Initially I had the idea to introduce new methods for that purpose:

``` csharp
public abstract class ScaffoldingTestsBase : FbTestsBase
{
	private readonly List<string> _tables = new();
	private readonly List<string> _domains = new();

	public ScaffoldingTestsBase()
		: base(FbServerType.Default, false, FbWireCrypt.Enabled, false)
	{ }


	protected async Task CreateTable(string name, params string[] fields)
	{
		using var command = Connection.CreateCommand();
		command.CommandText = $"create table {name} ({string.Join(", ", fields)})";
		await command.ExecuteNonQueryAsync();
		_tables.Add(name);
	}

	protected async Task CreateDomain(string name, string type)
	{
		using var command = Connection.CreateCommand();
		command.CommandText = $"create domain {name} as {type}";
		await command.ExecuteNonQueryAsync();
		_domains.Add(name);
	}

	[TearDown]
	public override async Task TearDown()
	{
		foreach (var table in _tables)
		{
			var command = Connection.CreateCommand();
			command.CommandText = $"drop table {table}";
			await command.ExecuteNonQueryAsync();
		}
		_tables.Clear();
		foreach (var domain in _domains)
		{
			var command = Connection.CreateCommand();
			command.CommandText = $"drop domain {domain}";
			await command.ExecuteNonQueryAsync();
		}
		_domains.Clear();
		await base.TearDown();
	}
}
```

But i am unsure if this is the right aproach here:
- These method would theoretically belong to `FbTestsBase` since creating tables or domains is not only required in scaffolding.
- The methods are not thread save which means when someone in furture starts to run tests with [Parallelizable] this will probably result in ungly errors.

Long story short: What is the best practice to remove database objects?

Comment (cincuranet) [reply src/FirebirdSql.EntityFrameworkCore.Firebird.Tests/Scaffolding/ScaffoldingTests.cs]: > The methods are not thread save which means when someone in furture starts to run tests with [Parallelizable] this will probably result in ungly errors.

Parallel (provider) testing with databases is can of worms. I have no plans to do parallel testing in foreseeable future, mostly because that would mean reworking lot of tests.

> Long story short: What is the best practice to remove database objects?

As you can see there's no consistent story (tests were created in different times with slightly varying setups). I think we can leave it as is, and when the time comes, the creation and cleanup will be unified.

**Comment by cincuranet** [comment]
Remove the `ScaffoldingTestsBase` and I'll merge.

**Comment by bgrauer-atacom** [comment]
Removed `ScaffoldingTestsBase`. Database objects are now not removed anymore.

**Comment by cincuranet** [comment]
Thanks for your contribution.
